crypto/provider_core.c

```diff
@@ -676,44 +744,134 @@ static int provider_init(OSSL_PROVIDER *prov, int flag_lock)
+        int i, max = sk_OSSL_PROVIDER_CHILD_CB_num(store->child_cbs);
+        OSSL_PROVIDER_CHILD_CB *child_cb;
 
-    if (--prov->activatecnt < 1)
         prov->flag_activated = 0;
+        for (i = 0; i < max; i++) {
+            child_cb = sk_OSSL_PROVIDER_CHILD_CB_value(store->child_cbs, i);
+            child_cb->remove_cb((OSSL_CORE_HANDLE *)prov, child_cb->cbdata);
+        }
+    }
 
     CRYPTO_THREAD_unlock(prov->flag_lock);
+    CRYPTO_THREAD_unlock(store->lock);
 
     /* We don't deinit here, that's done in ossl_provider_free() */
-    return 1;
+    return count;
 }
 
-static int provider_activate(OSSL_PROVIDER *prov, int flag_lock)
+/*
+ * Activate a provider.
+ * Return -1 on failure and the activation count on success
+ */
+static int provider_activate(OSSL_PROVIDER *prov, int lock, int upcalls)
 {
-    if (provider_init(prov, flag_lock)) {
-        if (flag_lock && !CRYPTO_THREAD_write_lock(prov->flag_lock))
+    int count = -1;
+
+    if (provider_init(prov, lock)) {
+        int ret = 1;
+        int i, max;
+        OSSL_PROVIDER_CHILD_CB *child_cb;
+        struct provider_store_st *store;
+
+        store = get_provider_store(prov->libctx);
+        if (store == NULL)
             return 0;
-        prov->activatecnt++;
-        prov->flag_activated = 1;
-        if (flag_lock)
-            CRYPTO_THREAD_unlock(prov->flag_lock);
 
-        return 1;
+        if (lock && !CRYPTO_THREAD_read_lock(store->lock))
+            return 0;
+
+        max = sk_OSSL_PROVIDER_CHILD_CB_num(store->child_cbs);
+        if (lock && !CRYPTO_THREAD_write_lock(prov->flag_lock))
+            return 0;
+
+#ifndef FIPS_MODULE
+        if (prov->ischild && upcalls)
+            ret = ossl_provider_up_ref_parent(prov, 1);
+#endif
+
+        if (ret) {
+            count = ++prov->activatecnt;
+            prov->flag_activated = 1;
+
+            if (prov->activatecnt == 1) {
+                for (i = 0; i < max; i++) {
+                    /*
+                     * This is newly activated (activatecnt == 1), so we need to
+                     * create child providers as necessary.
+                     */
+                    child_cb = sk_OSSL_PROVIDER_CHILD_CB_value(store->child_cbs,
+                                                               i);
+                    ret &= child_cb->create_cb((OSSL_CORE_HANDLE *)prov,
+                                               child_cb->cbdata);
+                }
+            }
+        }
+
+        if (lock) {
+            CRYPTO_THREAD_unlock(prov->flag_lock);
+            CRYPTO_THREAD_unlock(store->lock);
+        }
+        if (!ret)
+            return -1;
     }
 
-    return 0;
+    return count;
 }
 
-int ossl_provider_activate(OSSL_PROVIDER *prov, int retain_fallbacks)
+static int provider_flush_store_cache(const OSSL_PROVIDER *prov)
 {
+    struct provider_store_st *store;
+
+    if ((store = get_provider_store(prov->libctx)) == NULL)
+        return 0;
+
+    if (!store->freeing)
```

Comment:
You might need a rebase.  This needs to be done under a read lock to avoid different bit fields fighting over the memory location.